clouddriver-artifacts/src/main/java/com/netflix/spinnaker/clouddriver/artifacts/gitRepo/GitRepoFileSystem.java

```diff
@@ -0,0 +1,128 @@
+
+  private final GitRepoArtifactProviderProperties config;
+  private final Map<String, Lock> pathLocks = new ConcurrentHashMap<>();
+
+  public GitRepoFileSystem(GitRepoArtifactProviderProperties properties) {
+    this.config = properties;
+  }
+
+  public Path getLocalClonePath(String repoUrl, String branch) {
+    return Paths.get(CLONES_HOME.toString(), hashCoordinates(repoUrl, branch));
+  }
+
+  public boolean tryLock(String repoUrl, String branch, long time, TimeUnit unit)
+      throws InterruptedException {
+    String hash = hashCoordinates(repoUrl, branch);
+    Lock lock = pathLocks.computeIfAbsent(hash, k -> new ReentrantLock());
+    return lock.tryLock(time, unit);
+  }
+
+  public boolean tryLock(String cloneHashDir) {
+    Lock lock = pathLocks.computeIfAbsent(cloneHashDir, k -> new ReentrantLock());
+    return lock.tryLock();
+  }
+
+  public void unlock(String repoUrl, String branch) {
+    unlock(hashCoordinates(repoUrl, branch));
+  }
+
+  public void unlock(String cloneHashDir) {
+    if (!pathLocks.containsKey(cloneHashDir)) {
+      return;
+    }
+    Lock lock = pathLocks.remove(cloneHashDir);
+    lock.unlock();
+  }
+
+  public boolean canRetainClone() {
+    return config.getCloneRetentionMinutes() != 0 && hasFreeDisk();
+  }
+
+  private boolean hasFreeDisk() {
+    long currentSize = FileUtils.sizeOfDirectory(CLONES_HOME.toFile());
+    return currentSize >= 0 && currentSize < config.getCloneRetentionMaxBytes();
+  }
+
+  private String hashCoordinates(String repoUrl, String branch) {
+    String coordinates =
+        String.format(
+            "%s-%s",
+            Optional.ofNullable(repoUrl).orElse("unknownUrl"),
+            Optional.ofNullable(branch).orElse("defaultBranch"));
+    return Hashing.sha256().hashString(coordinates, Charset.defaultCharset()).toString();
+  }
+
+  @Scheduled(
+      fixedDelayString =
+          "${artifacts.git-repo.clone-retention-check-ms:"
+              + GitRepoArtifactProviderProperties.DEFAULT_CLONE_RETENTION_CHECK_MS
+              + "}")
+  private void deleteExpiredRepos() {
+    try {
+      if (!CLONES_HOME.toFile().exists() || config.getCloneRetentionMinutes() < 0) {
+        return;
+      }
+      File[] repos = CLONES_HOME.toFile().listFiles();
+      if (repos == null) {
+        return;
+      }
+      for (File r : repos) {
+        long ageMin = ((System.currentTimeMillis() - r.lastModified()) / 1000) / 60;
+        if (ageMin < config.getCloneRetentionMinutes()) {
+          continue;
+        }
+        if (!tryLock(r.getName())) {
+          // move on if the directory is locked by another thread, just wait for the next cycle
+          continue;
+        }
+        try {
+          log.info("Deleting expired git clone {}", r.getName());
+          FileUtils.forceDelete(r);
+        } finally {
+          unlock(r.getName());
+        }
+      }
+    } catch (IOException e) {
+      log.error("Error deleting expired git clones, ignoring", e);
```

Comment:
If we ignore, shouldn't it be a warning?

Reply:
I think `error` level is appropriate because if the expectation is that clones are deleted and that doesn't work, it could potentially lead to running out of disk and the pod getting killed.